### PR TITLE
Initial work to create a service for returning related documents for copilot purposes.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**

### CODE REVIEW

File: src/Features/Core/Portable/RelatedDocuments/AbstractRelatedDocumentsService.cs [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+using Microsoft.CodeAnalysis.Remote;
+
+namespace Microsoft.CodeAnalysis.RelatedDocuments;
+
+internal abstract class AbstractRelatedDocumentsService : IRelatedDocumentsService
+{
+    public async ValueTask GetRelatedDocumentIdsAsync(
+        Document document, int position, Func<DocumentId, CancellationToken, ValueTask> callbackAsync, CancellationToken cancellationToken)
+    {
+        var project = document.Project;
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/RelatedDocuments/AbstractRelatedDocumentsService.cs:17]: core abstract impl.  does nothing but attempt ot call out to oop to do this work sicne we're about to do a lot of binding.  would be subclassed in VB as well if we want to support that.

File: src/Features/Core/Portable/RelatedDocuments/IRelatedDocumentsService.cs [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+internal interface IRelatedDocumentsService : ILanguageService
+{
+    /// <summary>
+    /// Given an document, and an optional position in that document, streams a unique list of documents Ids that the
+    /// language think are "related".  It is up to the language to define what "related" means.  However, common
+    /// examples might be checking to see which symbols are used at that particular location and prioritizing documents
+    /// those symbols are defined in.
+    /// </summary>
+    ValueTask GetRelatedDocumentIdsAsync(Document document, int position, Func<DocumentId, CancellationToken, ValueTask> callbackAsync, CancellationToken cancellationToken);
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/RelatedDocuments/IRelatedDocumentsService.cs:20]: switched to callback-style (Vs IAsyncEnumerable).  it was just much easier to model, once i had all the OOP stuff hooked up.

File: eng/targets/Services.props [hunk 3]
```diff
@@ -13,38 +13,39 @@
   <ItemGroup>
     <ServiceHubService Include="Microsoft.VisualStudio.LanguageServices.AssetSynchronization" ClassName="Microsoft.CodeAnalysis.Remote.RemoteAssetSynchronizationService+Factory" />
     <ServiceHubService Include="Microsoft.VisualStudio.LanguageServices.AsynchronousOperationListener" ClassName="Microsoft.CodeAnalysis.Remote.RemoteAsynchronousOperationListenerService+Factory" Audience="AllClientsIncludingGuests"/>
+    <ServiceHubService Include="Microsoft.VisualStudio.LanguageServices.CodeLensReferences" ClassName="Microsoft.CodeAnalysis.Remote.RemoteCodeLensReferencesService+Factory" />
```
Comment (CyrusNajmabadi) [inline eng/targets/Services.props:16]: just sorted.

File: eng/targets/Services.props [hunk 4]
```diff
@@ -13,38 +13,39 @@
...
-    <ServiceHubService Include="Microsoft.VisualStudio.LanguageServices.CodeLensReferences" ClassName="Microsoft.CodeAnalysis.Remote.RemoteCodeLensReferencesService+Factory" />
-    <ServiceHubService Include="Microsoft.VisualStudio.LanguageServices.DesignerAttributeDiscovery" ClassName="Microsoft.CodeAnalysis.Remote.RemoteDesignerAttributeDiscoveryService+Factory" />
+    <ServiceHubService Include="Microsoft.VisualStudio.LanguageServices.ProcessTelemetry" ClassName="Microsoft.CodeAnalysis.Remote.RemoteProcessTelemetryService+Factory" />
     <ServiceHubService Include="Microsoft.VisualStudio.LanguageServices.ProjectTelemetry" ClassName="Microsoft.CodeAnalysis.Remote.RemoteProjectTelemetryService+Factory" />
-    <ServiceHubService Include="Microsoft.VisualStudio.LanguageServices.LegacySolutionEventsAggregation" ClassName="Microsoft.CodeAnalysis.Remote.RemoteLegacySolutionEventsAggregationService+Factory" />
+    <ServiceHubService Include="Microsoft.VisualStudio.LanguageServices.RelatedDocuments" ClassName="Microsoft.CodeAnalysis.Remote.RemoteRelatedDocumentsService+Factory" />
```
Comment (CyrusNajmabadi) [inline eng/targets/Services.props:37]: this is new.

File: src/Features/CSharp/Portable/RelatedDocuments/CSharpRelatedDocumentsService.cs [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.RelatedDocuments;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+
+namespace Microsoft.CodeAnalysis.CSharp.RelatedDocuments;
+
+[ExportLanguageService(typeof(IRelatedDocumentsService), LanguageNames.CSharp), Shared]
+[method: ImportingConstructor]
+[method: Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+internal sealed class CSharpRelatedDocumentsService() : AbstractRelatedDocumentsService<
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/RelatedDocuments/CSharpRelatedDocumentsService.cs:20]: language specific logic.  for now, the lang specific logic is just iterating and finding which nodes to bind.  we bind Name nodes (X, X.Y, `X<Y>`, etc.), and member access nodes with a left dotter name (like the `X.Y` in `X.Y.Z()`).

File: src/Workspaces/Remote/ServiceHub/Services/RelatedDocuments/RemoteRelatedDocumentsService.cs [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+    private readonly RemoteCallback<IRemoteRelatedDocumentsService.ICallback> _callback = callback;
+
+    public ValueTask GetRelatedDocumentIdsAsync(
+        Checksum solutionChecksum,
+        DocumentId documentId,
+        int position,
+        RemoteServiceCallbackId callbackId,
+        CancellationToken cancellationToken)
+    {
+        return RunServiceAsync(solutionChecksum, async solution =>
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Remote/ServiceHub/Services/RelatedDocuments/RemoteRelatedDocumentsService.cs:33]: oops service is trivial, it just receives the inputs, gets the service on the oop side, and calls into it.

**Comment by CyrusNajmabadi** [comment]
@dibarbet  @genlu ptal.  The lsp handler PR comes after this one.

**Comment by CyrusNajmabadi** [comment]
NOte: as this is green, i'd like to roll any non-critical changes into: https://github.com/dotnet/roslyn/pull/74918

File: src/Features/CSharp/Portable/RelatedDocuments/CSharpRelatedDocumentsService.cs [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+namespace Microsoft.CodeAnalysis.CSharp.RelatedDocuments;
+
+[ExportLanguageService(typeof(IRelatedDocumentsService), LanguageNames.CSharp), Shared]
+[method: ImportingConstructor]
+[method: Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+internal sealed class CSharpRelatedDocumentsService() : AbstractRelatedDocumentsService
+{
+    protected override async ValueTask GetRelatedDocumentIdsInCurrentProcessAsync(
+        Document document,
+        int position,
```
Comment (genlu) [inline src/Features/CSharp/Portable/RelatedDocuments/CSharpRelatedDocumentsService.cs]: position needs to be optional.

Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/RelatedDocuments/CSharpRelatedDocumentsService.cs]: it is.  you can just pass 0.

File: src/Features/CSharp/Portable/RelatedDocuments/CSharpRelatedDocumentsService.cs [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+            {
+                if (current is MemberAccessExpressionSyntax memberAccess)
+                {
+                    // Could be a static member access off of a type name.  Check the left side, and if it's just a
+                    // dotted name, return that.
+
+                    if (IsPossibleTypeName(memberAccess.Expression, out var nameToken))
+                    {
+                        // Something like `X.Y.Z` where `X.Y` is a type name.  Bind X.Y
+                        yield return (memberAccess.Expression, nameToken);
```
Comment (genlu) [inline src/Features/CSharp/Portable/RelatedDocuments/CSharpRelatedDocumentsService.cs]: For `X.Y.Z`, memberAccess.Expression is `X.Y`, but the nameToken returned would be `X`, right? Shouldn't it be `Y`?

File: src/Features/CSharp/Portable/RelatedDocuments/CSharpRelatedDocumentsService.cs [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+        var seenTypeNames = new ConcurrentSet<string>();
+
+        // Bind as much as we can in parallel to get the results as quickly as possible.  We call into the
+        // ProducerConsumer overload that will batch up values into arrays, and call into the callback with them.  While
+        // that callback is executing, the ProducerConsumer continues to run, computing more results and getting them
+        // ready for when that call returns.  This approach ensures that we're not pausing while we're reporting the
+        // results to whatever client is calling into us.
+        await ProducerConsumer<DocumentId>.RunParallelAsync(
+            // Order the nodes by the distance from the requested position.
+            IteratePotentialTypeNodes(root).OrderBy(t => t.expression.SpanStart - position),
```
Comment (genlu) [inline src/Features/CSharp/Portable/RelatedDocuments/CSharpRelatedDocumentsService.cs]: > t.expression.SpanStart - position [](http://example.com/codeflow?start=56&length=34)

Should this be the absolute value of `t.expression.SpanStart - position` ?

**Review by genlu — APPROVED** [review]
This is a solid starting point. I have a couple of question about the code, other than those LGTM. Feel free to address in another PR

One concern is the perf/behavior on really large files. We might consider a different grouping strategy like

1. (if a position is provided) find types referenced within enclosing member 
2. types referenced in other members' signatures
3. types referenced inside other members

We could always try to return everything from first group in the first callback, so consumer of the API would always get the most relevant ones and have an opportunity to decide if they want to wait for more. Thoughts?

File: src/Features/CSharp/Portable/RelatedDocuments/CSharpRelatedDocumentsService.cs (hunk 7 above)
Comment (genlu) [reply src/Features/CSharp/Portable/RelatedDocuments/CSharpRelatedDocumentsService.cs]: Consider using -1 for this purpose. I think 0 is a valid position in file, so potentially we could return a different list of related files for "position 0" compare to "the entire file"

File: src/Features/CSharp/Portable/RelatedDocuments/CSharpRelatedDocumentsService.cs (hunk 8 above)
Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/RelatedDocuments/CSharpRelatedDocumentsService.cs]: not exactly.  we have `X.Y.Z`.  But we pass `X.Y` Into IsPossibleTypeName, and if it succeeds, we return `Y`.

File: src/Features/CSharp/Portable/RelatedDocuments/CSharpRelatedDocumentsService.cs (hunk 9 above)
Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/RelatedDocuments/CSharpRelatedDocumentsService.cs]: yup.  That would make more sense!

**Comment by CyrusNajmabadi** [comment]
> We could always try to return everything from first group in the first callback, so consumer of the API would always get the most relevant ones and have an opportunity to decide if they want to wait for more. Thoughts?

Totally.  Though note that the design of this (including with LSP) is to be entirely streaming based.   The service just keeps calling the callback as it finds results, and teh client just pulls as long as it wants.

**Comment by CyrusNajmabadi** [comment]
> We might consider a different grouping strategy like ...

Yup.  All those ideas are great.  This was meant more as a starting position.  The code that iterates the nodes can be greatly tweaked as we go forward.

**Pull request merged**